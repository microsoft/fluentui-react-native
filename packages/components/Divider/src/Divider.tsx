/** @jsxRuntime classic */
/** @jsx withSlots */
/** @jsxFrag */
import React from 'react';
import { View } from 'react-native';
import type { ViewProps } from 'react-native';

import { withSlots, compressible, useSlot, useFluentTheme, patchTokens } from '@fluentui-react-native/framework';
import type { UseTokens } from '@fluentui-react-native/framework';
import { IconV1 as Icon } from '@fluentui-react-native/icon';
import type { IconPropsV1 as IconProps } from '@fluentui-react-native/icon';
import { TextV1 as Text } from '@fluentui-react-native/text';
import type { TextProps } from '@fluentui-react-native/text';

import { colorsFromAppearance, getBeforeLineStyle, getRootStyle, useDividerSlotProps } from './Divider.styling';
import { dividerName } from './Divider.types';
import type { DividerProps, DividerTokens } from './Divider.types';
import { useDividerTokens } from './DividerTokens';

export const Divider = compressible<DividerProps, DividerTokens>((props: DividerProps, useTokens: UseTokens<DividerTokens>) => {
  // Set default values for props
  props = {
    alignContent: 'center',
    appearance: 'default',
    insetSize: 0,
    vertical: false,
    ...props,
  };

  const theme = useFluentTheme();
  let [tokens, cache] = useTokens(theme);

  // We want to patch our initial tokens from useTokens() with the token values below because they are dependent on the above props.
  // With compressible, we can also access our tokens at the final rendering stage, which we will need to correctly set our divider height.
  [tokens, cache] = patchTokens(tokens, cache, {
    flexAfter: props.alignContent === 'end' ? 0 : 1,
    flexBefore: props.alignContent === 'start' ? 0 : 1,
    ...colorsFromAppearance(props.appearance, tokens, theme),
  });

  // get slot props from these tokens
  const { rootProps, beforeLineProps, afterLineProps, wrapperProps, textProps, iconProps } = useDividerSlotProps(props, tokens, theme);

  // build slots
  const RootSlot = useSlot<ViewProps>(View, rootProps);
  const BeforeLineSlot = useSlot<ViewProps>(View, beforeLineProps);
  const AfterLineSlot = useSlot<ViewProps>(View, afterLineProps);
  const WrapperSlot = useSlot<ViewProps>(View, wrapperProps);
  const TextSlot = useSlot<TextProps>(Text, textProps);
  const IconSlot = useSlot<IconProps>(Icon, iconProps);

  return (final: DividerProps, ...children: React.ReactNode[]) => {
    final = { ...props, ...final };
    // change root style if there is a text child
    let textContent: string;
    React.Children.forEach(children, (child) => {
      if (typeof child === 'string') {
        textContent = child;
      }
    });

    const hasContent = textContent !== undefined || props.icon !== undefined;

    // This style must be set here because we need to know if text content is passed in the final render to set the height correctly
    let finalRootProps = rootProps;
    if (!tokens.minHeight) {
      finalRootProps = { ...rootProps, style: getRootStyle(rootProps.style, final.vertical, hasContent) };
    }

    // We patch the beforeLine styling if no content is passed because it should always take up the full width / height of the root container (afterLine is not rendered).
    const finalBeforeLineProps = { ...beforeLineProps, style: getBeforeLineStyle(beforeLineProps.style, hasContent) };

    return (
      <RootSlot {...finalRootProps}>
        <BeforeLineSlot {...finalBeforeLineProps} />
        {hasContent && (
          <>
            <WrapperSlot>
              {textContent && <TextSlot>{textContent}</TextSlot>}
              {props.icon && !textContent && <IconSlot />}
            </WrapperSlot>
            <AfterLineSlot />
          </>
        )}
      </RootSlot>
    );
  };
}, useDividerTokens);
Divider.displayName = dividerName;
