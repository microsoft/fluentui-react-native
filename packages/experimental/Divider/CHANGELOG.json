{
  "name": "@fluentui-react-native/divider",
  "entries": [
    {
      "date": "Fri, 24 Feb 2023 20:39:57 GMT",
      "tag": "@fluentui-react-native/divider_v0.3.1",
      "version": "0.3.1",
      "comments": {
        "patch": [
          {
            "author": "78454019+lyzhan7@users.noreply.github.com",
            "package": "@fluentui-react-native/divider",
            "commit": "743bd1507af336a62946361f3fe3c800ae5bd4c2",
            "comment": "Order imports"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/divider",
            "comment": "Bump @fluentui-react-native/framework to v0.9.3",
            "commit": "743bd1507af336a62946361f3fe3c800ae5bd4c2"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/divider",
            "comment": "Bump @fluentui-react-native/icon to v0.17.14",
            "commit": "743bd1507af336a62946361f3fe3c800ae5bd4c2"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/divider",
            "comment": "Bump @fluentui-react-native/text to v0.19.22",
            "commit": "743bd1507af336a62946361f3fe3c800ae5bd4c2"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/divider",
            "comment": "Bump @fluentui-react-native/tokens to v0.20.10",
            "commit": "743bd1507af336a62946361f3fe3c800ae5bd4c2"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/divider",
            "comment": "Bump @fluentui-react-native/theme-tokens to v0.24.3",
            "commit": "743bd1507af336a62946361f3fe3c800ae5bd4c2"
          }
        ]
      }
    },
    {
      "date": "Fri, 24 Feb 2023 02:22:40 GMT",
      "tag": "@fluentui-react-native/divider_v0.3.0",
      "version": "0.3.0",
      "comments": {
        "minor": [
          {
            "author": "winlarry@microsoft.com",
            "package": "@fluentui-react-native/divider",
            "commit": "a266ece54bfecbd7edd912a26f46dae655bf48ef",
            "comment": "Added styling tokens for styling Divider text"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/divider",
            "comment": "Bump @fluentui-react-native/framework to v0.9.2",
            "commit": "a266ece54bfecbd7edd912a26f46dae655bf48ef"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/divider",
            "comment": "Bump @fluentui-react-native/icon to v0.17.13",
            "commit": "a266ece54bfecbd7edd912a26f46dae655bf48ef"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/divider",
            "comment": "Bump @fluentui-react-native/text to v0.19.21",
            "commit": "a266ece54bfecbd7edd912a26f46dae655bf48ef"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/divider",
            "comment": "Bump @fluentui-react-native/tokens to v0.20.9",
            "commit": "a266ece54bfecbd7edd912a26f46dae655bf48ef"
          }
        ]
      }
    },
    {
      "date": "Wed, 22 Feb 2023 18:07:14 GMT",
      "tag": "@fluentui-react-native/divider_v0.2.4",
      "version": "0.2.4",
      "comments": {
        "patch": [
          {
            "author": "winlarry@microsoft.com",
            "package": "@fluentui-react-native/divider",
            "commit": "ffe9122b5e6f7d93f7efbff5a77fa8efd15db6f3",
            "comment": "Add more snapshot testing examples"
          }
        ]
      }
    },
    {
      "date": "Fri, 17 Feb 2023 01:34:44 GMT",
      "tag": "@fluentui-react-native/divider_v0.2.3",
      "version": "0.2.3",
      "comments": {
        "patch": [
          {
            "author": "winlarry@microsoft.com",
            "package": "@fluentui-react-native/divider",
            "commit": "0340ca1e7dff01aa00d6a7240fc7b06c9787c3e8",
            "comment": "Update SPEC.md documentation for Divider"
          }
        ]
      }
    },
    {
      "date": "Wed, 15 Feb 2023 18:30:10 GMT",
      "tag": "@fluentui-react-native/divider_v0.2.2",
      "version": "0.2.2",
      "comments": {
        "patch": [
          {
            "author": "rohanpd.work@gmail.com",
            "package": "@fluentui-react-native/divider",
            "commit": "ba4cab7eef8759aee74cb0d4ca31d2b8360e3724",
            "comment": "impl mobile divider"
          }
        ]
      }
    },
    {
      "date": "Wed, 15 Feb 2023 00:51:10 GMT",
      "tag": "@fluentui-react-native/divider_v0.2.1",
      "version": "0.2.1",
      "comments": {
        "patch": [
          {
            "author": "winlarry@microsoft.com",
            "package": "@fluentui-react-native/divider",
            "commit": "4ea4d97ea0fa8cea2a682c43823c0e2c850e5066",
            "comment": "Appearance prop correctly affects color"
          }
        ]
      }
    },
    {
      "date": "Sat, 11 Feb 2023 01:32:30 GMT",
      "tag": "@fluentui-react-native/divider_v0.2.0",
      "version": "0.2.0",
      "comments": {
        "minor": [
          {
            "author": "winlarry@microsoft.com",
            "package": "@fluentui-react-native/divider",
            "commit": "271dc7378669e31e355a4ca8f131a96872d8049a",
            "comment": "Implement win32 DividerV1 component"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/divider",
            "comment": "Bump @fluentui-react-native/framework to v0.9.1",
            "commit": "271dc7378669e31e355a4ca8f131a96872d8049a"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/divider",
            "comment": "Bump @fluentui-react-native/icon to v0.17.12",
            "commit": "271dc7378669e31e355a4ca8f131a96872d8049a"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/divider",
            "comment": "Bump @fluentui-react-native/text to v0.19.20",
            "commit": "271dc7378669e31e355a4ca8f131a96872d8049a"
          },
          {
            "author": "beachball",
            "package": "@fluentui-react-native/divider",
            "comment": "Bump @fluentui-react-native/tokens to v0.20.8",
            "commit": "271dc7378669e31e355a4ca8f131a96872d8049a"
          }
        ]
      }
    }
  ]
}
