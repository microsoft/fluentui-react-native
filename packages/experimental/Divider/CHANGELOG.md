# Change Log - @fluentui-react-native/divider

This log was last generated on Fri, 24 Feb 2023 20:39:57 GMT and should not be manually modified.

<!-- Start content -->

## 0.3.1

Fri, 24 Feb 2023 20:39:57 GMT

### Patches

- Order imports (78454019+lyzhan7@users.noreply.github.com)
- Bump @fluentui-react-native/framework to v0.9.3
- Bump @fluentui-react-native/icon to v0.17.14
- Bump @fluentui-react-native/text to v0.19.22
- Bump @fluentui-react-native/tokens to v0.20.10
- Bump @fluentui-react-native/theme-tokens to v0.24.3

## 0.3.0

Fri, 24 Feb 2023 02:22:40 GMT

### Minor changes

- Added styling tokens for styling Divider text (winlarry@microsoft.com)
- Bump @fluentui-react-native/framework to v0.9.2
- Bump @fluentui-react-native/icon to v0.17.13
- Bump @fluentui-react-native/text to v0.19.21
- Bump @fluentui-react-native/tokens to v0.20.9

## 0.2.4

Wed, 22 Feb 2023 18:07:14 GMT

### Patches

- Add more snapshot testing examples (winlarry@microsoft.com)

## 0.2.3

Fri, 17 Feb 2023 01:34:44 GMT

### Patches

- Update SPEC.md documentation for Divider (winlarry@microsoft.com)

## 0.2.2

Wed, 15 Feb 2023 18:30:10 GMT

### Patches

- impl mobile divider (rohanpd.work@gmail.com)

## 0.2.1

Wed, 15 Feb 2023 00:51:10 GMT

### Patches

- Appearance prop correctly affects color (winlarry@microsoft.com)

## 0.2.0

Sat, 11 Feb 2023 01:32:30 GMT

### Minor changes

- Implement win32 DividerV1 component (winlarry@microsoft.com)
- Bump @fluentui-react-native/framework to v0.9.1
- Bump @fluentui-react-native/icon to v0.17.12
- Bump @fluentui-react-native/text to v0.19.20
- Bump @fluentui-react-native/tokens to v0.20.8
