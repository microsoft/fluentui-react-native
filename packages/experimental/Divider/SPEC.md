# Divider

## Background

The `Divider` component represents a visual separator on a given page. On desktop platforms, a `Divider` can be horizontal or vertical, and it can contain content, text or an icon, for aesthetic purposes.

Note:

For mobile platforms, only horizontal `Divider` is supported.

Content, text or an icon are also not supported on mobile platforms.

## Sample Code

### Basic Divider

```tsx
<>
  <Divider /> // This is a horizontal divider with no content.
  <Divider insetSize={16} /> // Divider with a defined inset
  <Divider vertical /> // Vertical divider
  <Divider appearance="brand" /> // Divider with a set appearance
</>
```

### Divider With Text

```tsx
<>
  <Divider>Lorem ipsum</Divider> // Center-aligned horizontal divider
  <Divider alignContent="start">Dolor sit</Divider> // Start-aligned horizontal divider
  <Divider alignContent="end" vertical>
    Amet consectetur
  </Divider>{' '}
  // End-aligned vertical divider
</>
```

### Divider With Icon

```tsx
const svgIconProps = {...} // insert svg icon props here
const fontIconProps = {...} // insert font icon props here

<>
  <Divider icon={{ svgSource: svgIconProps }} /> // With svg icon
  <Divider icon={{ fontSource: fontIconProps }} /> // With font icon
</>
```

## Variants

### Vertical

The `Divider` will be rendered as a vertical line, take up a minimum height of 84px, and not take up the entire row if the `vertical` prop is set to `true`.

### With Content

If text as passed as a child, or the icon prop is set, the `Divider` will render the content between the dividing line. It will also take up more width and height depending on the `Divider` orientation.

#### Alignment

Users can control where content appears in the `Divider`.

- If `alignContent` is equal to "start", the content appears towards the left of the `Divider` if it's horizontal or towards the top if vertical.
- If `alignContent` is equal to "end", the content appears towards the right of the `Divider` if it's horizontal or towards the bottom if vertical.
- If `alignContent` is equal to "center" or not set, the content appears exactly in the middle of the dividing line.

## API

### Slots

The `Divider` uses 6 total slots:

- `root` [View] The outer container of the component
- `beforeLine` [View] The dividing line rendered before passed content. If no content is passed, this is the only slot rendered.
- `afterLine` [View] The dividing line rendered after passed content. Not rendered if no content is passed.
- `wrapper` [View] If content is passed, this is the outer container of the content between `beforeLine` and `afterLine` which provides the correct padding for the content.
- `text` [TextV1] If text is passed as a child, this renders inside `wrapper`.
- `icon` [IconV1] If the `icon` prop is set, an icon will render inside `wrapper`. If text is also passed while the `icon` prop is set, then this slot will not render.

Because the Divider is built using the `compressible` framework, these slots are currently not able to be modified.

### Props

Below is the set of props `Divider` supports.

```tsx
export interface DividerProps {
  /**
   * If a text or icon is passed, this dictates where content appears in the divider: at the start, centered, or towards the end.
   * @default 'center'
   * Note: This prop is not supported on mobile platforms(Android & iOS).
   */
  alignContent?: DividerAlignment;
  /**
   * If no color tokens are set, the divider and its content are colored using different theme tokens depending on the value of this prop.
   * @default 'default'
   * Note: This prop is not supported on mobile platforms(Android & iOS).
   */
  appearance?: DividerAppearance;
  /**
   * Pass an icon source to render an icon as content in the divider. Will not render if text is passed via children.
   * Note: This prop is not supported on mobile platforms(Android & iOS).
   */
  icon?: IconProps;
  /**
   * The size of the Divider inset - the margin before the start and after the end of the divider.
   * @default 0
   * Note : For mobile platforms, the insetSize prop is only applied to start of the component.
   */
  insetSize?: DividerInsetSize;
  /**
   * Whether the divider is rendered as a horizontal line or a vertical line.
   * @default false
   * Note: This prop is not supported on mobile platforms(Android & iOS).
   */
  vertical?: boolean;
}
```

### Styling Tokens

Tokens can be used to customize the styling of the control by using the customize function on the `Divider`. For more information on using the customize API, please see [this page](https://github.com/microsoft/fluentui-react-native/blob/main/packages/framework/composition/README.md). The `Divider` has the following tokens:

```tsx
export interface DividerTokens extends LayoutTokens, Omit<FontTokens, 'fontDynamicTypeRamp' | 'fontMaximumSize'> {
  /**
   * The color of the content passed into the divider.
   */
  contentColor?: ColorValue;
  /**
   * The padding of divider content between the start and end lines.
   * @default 12
   */
  contentPadding?: string | number;
  /**
   * The flex value of the line after content, set to 0 if `alignContent` = `end`.
   * @default 1
   */
  flexAfter?: number;
  /**
   * The flex value of the line before content, set to 0 if `alignContent` = `start`.
   * @default 1
   */
  flexBefore?: number;
  /**
   * Color of the divider lines.
   */
  lineColor?: ColorValue;
  /**
   * The minimum size of a line shrunken by a non-centered divider.
   * @default 8
   */
  minLineSize?: number;
  /**
   * The thickness of the Divider lines
   * @default 1
   */
  thickness?: number;
}
```

In addition to the tokens listed here, Divider also has tokens to control text and layout styling.

#### Text Tokens

If text is passed into the divider, the following tokens can be used to add styling to the text.

```tsx
// not an actual type - these are the properties DividerTokens inherit
type DividerFontTokens = {
  /**
   * Change the font family of the passed text
   */
  fontFamily?: string;
  /**
   * Increase or decrease the size of the passed text.
   */
  fontSize?: keyof Typography['sizes'] | TextStyle['fontSize'];
  /**
   * Make text bolder or thinner.
   */
  fontWeight?: keyof Typography['weights'] | TextStyle['fontWeight'];
  /**
   * Change the height of each line of text rendered.
   */
  fontLineHeight?: number;
  /**
   * Add spacing between each character within the text string.
   */
  fontLetterSpacing?: number;
  /**
   * Use to make text italic.
   */
  fontStyle?: 'normal' | 'italic';
  /**
   * Render an underline, a strikethrough-line, or both with the text.
   */
  textDecorationLine?: 'normal' | 'underline' | 'line-through' | 'underline line-through';
  /**
   * Pass a variant type in to use a preset set of Fluent text styles for different purposes (e.g. styling a header, a caption).
   *
   * If you specify a variant while modifying a text style (e.g. passing a variant while also passing a custom fontSize), your styles will take precedence.
   */
  variant?: Variants;
};
```

#### Layout Tokens

To apply margins / padding to the divider, adjust these tokens.

```tsx
// not an actual type - these are the properties DividerTokens inherit
type DividerLayoutTokens = {
  /**
   * Adjust the tokens below to set ceilings and floors on the divider width and height.
   */
  minWidth?: string | number;
  maxWidth?: string | number;
  /**
   * For minHeight, there are different default values:
   * @default 0 for horizontal dividers
   * @default 20 for vertical dividers without children
   * @default 84 for vertical dividers with children
   */
  minHeight?: string | number;
  maxHeight?: string | number;
  /**
   * Adjust the tokens below to set custom spacing for the divider (if inset is not enough).
   */
  padding?: string | number;
  paddingHorizontal?: string | number;
  paddingVertical?: string | number;
  paddingStart?: string | number;
  paddingEnd?: string | number;
};
```
