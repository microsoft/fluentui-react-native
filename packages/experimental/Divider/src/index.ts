export { Divider } from './Divider';
export { DividerTokens, DividerProps, DividerSlotProps, DividerType } from './Divider.types';
