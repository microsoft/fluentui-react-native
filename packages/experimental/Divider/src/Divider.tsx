/** @jsx withSlots */
/** @jsxFrag */
import { View, ViewProps } from 'react-native';
import { dividerName, DividerProps, DividerTokens } from './Divider.types';
import { withSlots, compressible, UseTokens, useSlot, useFluentTheme, patchTokens, mergeStyles } from '@fluentui-react-native/framework';
import { Text, TextProps } from '@fluentui-react-native/text';
import { IconV1, IconPropsV1 } from '@fluentui-react-native/icon';
import React from 'react';
import { useDividerTokens } from './DividerTokens';
import { globalTokens } from '@fluentui-react-native/theme-tokens';
import { colorsFromAppearance, getDividerSlotProps } from './Divider.styling';

export const Divider = compressible<DividerProps, DividerTokens>((props: DividerProps, useTokens: UseTokens<DividerTokens>) => {
  const { alignContent = 'center', appearance = 'default', color, icon, insetSize = 0, vertical = false } = props;

  const theme = useFluentTheme();
  let [tokens, cache] = useTokens(theme);

  // call patch function to manually change token values from props (this is done by the lookup function in compose components)
  [tokens, cache] = patchTokens(tokens, cache, {
    color,
    insetSize,
    vertical,
    ...colorsFromAppearance(appearance, theme),
    contentPadding: globalTokens.size80,
    flexAfter: alignContent === 'end' ? 0 : 1,
    flexBefore: alignContent === 'start' ? 0 : 1,
    minHeight: props.vertical ? globalTokens.size240 : 0,
  });

  // get slot props from these tokens
  const {
    root: rootProps,
    beforeLine: beforeLineProps,
    afterLine: afterLineProps,
    wrapper: wrapperProps,
    text: textProps,
    icon: iconProps,
  } = React.useMemo(() => getDividerSlotProps(tokens, theme, icon), [tokens, theme, icon]);

  // build slots
  const RootSlot = useSlot<ViewProps>(View, rootProps);
  const BeforeLineSlot = useSlot<ViewProps>(View, beforeLineProps);
  const AfterLineSlot = useSlot<ViewProps>(View, afterLineProps);
  const WrapperSlot = useSlot<ViewProps>(View, wrapperProps);
  const TextSlot = useSlot<TextProps>(Text, textProps);
  const IconSlot = useSlot<IconPropsV1>(IconV1, iconProps);

  return (final: DividerProps, ...children: React.ReactNode[]) => {
    props = { ...props, ...final };
    // change root style if there is a text child
    let textContent: string;
    React.Children.forEach(children, (child) => {
      if (typeof child === 'string') {
        textContent = child;
      }
    });

    const hasContent = textContent !== undefined || props.icon !== undefined;

    // This style must be set here because we need to know if text content is passed in the final render
    const mergedProps = {
      ...rootProps,
      style: mergeStyles(rootProps, props.vertical && textContent ? { minHeight: 84 } : {}),
    };

    return (
      <RootSlot {...mergedProps}>
        <BeforeLineSlot />
        {hasContent && (
          <>
            <WrapperSlot>
              {textContent && <TextSlot>{textContent}</TextSlot>}
              {props.icon && !textContent && <IconSlot />}
            </WrapperSlot>
            <AfterLineSlot />
          </>
        )}
      </RootSlot>
    );
  };
}, useDividerTokens);
Divider.displayName = dividerName;
