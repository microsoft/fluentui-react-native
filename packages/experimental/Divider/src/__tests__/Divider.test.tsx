import * as React from 'react';

import { checkReRender } from '@fluentui-react-native/test-tools';
import * as renderer from 'react-test-renderer';

import { Divider } from '../Divider';
import type { DividerProps } from '../Divider.types';

describe('Divider component tests', () => {
  it('Divider default', () => {
    const tree = renderer.create(<Divider />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Vertical Divider', () => {
    const tree = renderer.create(<Divider vertical />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Subtle Divider', () => {
    const tree = renderer.create(<Divider appearance="subtle" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Branded Divider', () => {
    const tree = renderer.create(<Divider appearance="brand" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Strong Divider', () => {
    const tree = renderer.create(<Divider appearance="strong" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Horizontal Divider with Inset', () => {
    const tree = renderer.create(<Divider insetSize={16} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Vertical Divider with Inset', () => {
    const tree = renderer.create(<Divider vertical insetSize={16} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Custom Divider', () => {
    const CustomDivider = Divider.customize({ thickness: 3, lineColor: 'red', contentColor: 'blue' });
    const tree = renderer.create(<CustomDivider />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Divider with text', () => {
    const tree = renderer.create(<Divider>Lorem Ipsum</Divider>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Divider with icon', () => {
    const tree = renderer.create(<Divider icon={{ fontSource: { fontFamily: 'Arial', codepoint: 0x2663, fontSize: 32 } }} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Divider with all props + tokens set', () => {
    const CustomDivider = Divider.customize({
      contentColor: 'red',
      contentPadding: 10,
      flexAfter: 2,
      flexBefore: 1,
      lineColor: 'blue',
      minLineSize: 10,
      thickness: 2,
      minHeight: 40,
      minWidth: 10,
      maxHeight: 200,
      maxWidth: 200,
      padding: 10,
      paddingVertical: 10,
      paddingHorizontal: 10,
      paddingStart: 10,
      paddingEnd: 10,
      fontFamily: 'serif',
      fontSize: 10,
      fontWeight: '600',
      fontLineHeight: 2,
      fontLetterSpacing: 0.1,
      fontStyle: 'italic',
      textDecorationLine: 'line-through',
      variant: 'body1',
    });
    const props: DividerProps = {
      alignContent: 'start',
      appearance: 'strong',
      icon: { fontSource: { fontFamily: 'Arial', codepoint: 0x2663, fontSize: 32 } },
      insetSize: 16,
      vertical: true,
    };
    const tree = renderer.create(<CustomDivider {...props}>Hello</CustomDivider>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Divider re-renders correctly', () => {
    checkReRender(() => <Divider />, 2);
  });
});
