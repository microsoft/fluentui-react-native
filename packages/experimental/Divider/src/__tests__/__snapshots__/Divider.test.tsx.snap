// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider component tests Branded Divider 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 0,
      "minWidth": 0,
      "paddingHorizontal": 0,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "#0078d4",
        "borderStyle": "solid",
        "borderTopWidth": 1,
        "flex": 1,
        "flexBasis": 8,
        "minWidth": 8,
      }
    }
  />
</View>
`;

exports[`Divider component tests Custom Divider 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 0,
      "minWidth": 0,
      "paddingHorizontal": 0,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "red",
        "borderStyle": "solid",
        "borderTopWidth": 3,
        "flex": 1,
        "flexBasis": 8,
        "minWidth": 8,
      }
    }
  />
</View>
`;

exports[`Divider component tests Divider default 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 0,
      "minWidth": 0,
      "paddingHorizontal": 0,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "#e0e0e0",
        "borderStyle": "solid",
        "borderTopWidth": 1,
        "flex": 1,
        "flexBasis": 8,
        "minWidth": 8,
      }
    }
  />
</View>
`;

exports[`Divider component tests Divider with icon 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 0,
      "minWidth": 0,
      "paddingHorizontal": 0,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "#e0e0e0",
        "borderStyle": "solid",
        "borderTopWidth": 1,
        "flex": 1,
        "flexBasis": 8,
        "minWidth": 8,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 0,
        "paddingHorizontal": 12,
      }
    }
  >
    <Text
      accessible={true}
      style={
        Object {
          "color": "#424242",
          "fontFamily": "Arial",
          "fontSize": 32,
        }
      }
    >
      â™£
    </Text>
  </View>
  <View
    style={
      Object {
        "borderColor": "#e0e0e0",
        "borderStyle": "solid",
        "borderTopWidth": 1,
        "flex": 1,
        "flexBasis": 8,
        "minWidth": 8,
      }
    }
  />
</View>
`;

exports[`Divider component tests Divider with text 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 0,
      "minWidth": 0,
      "paddingHorizontal": 0,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "#e0e0e0",
        "borderStyle": "solid",
        "borderTopWidth": 1,
        "flex": 1,
        "flexBasis": 8,
        "minWidth": 8,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 0,
        "paddingHorizontal": 12,
      }
    }
  >
    <Text
      ellipsizeMode="tail"
      numberOfLines={0}
      onAccessibilityTap={[Function]}
      style={
        Object {
          "color": "#424242",
          "fontFamily": "Segoe UI",
          "fontSize": 12,
          "fontWeight": "400",
          "margin": 0,
          "textAlign": "center",
        }
      }
    >
      Lorem Ipsum
    </Text>
  </View>
  <View
    style={
      Object {
        "borderColor": "#e0e0e0",
        "borderStyle": "solid",
        "borderTopWidth": 1,
        "flex": 1,
        "flexBasis": 8,
        "minWidth": 8,
      }
    }
  />
</View>
`;

exports[`Divider component tests Horizontal Divider with Inset 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 0,
      "minWidth": 0,
      "paddingHorizontal": 16,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "#e0e0e0",
        "borderStyle": "solid",
        "borderTopWidth": 1,
        "flex": 1,
        "flexBasis": 8,
        "minWidth": 8,
      }
    }
  />
</View>
`;

exports[`Divider component tests Strong Divider 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 0,
      "minWidth": 0,
      "paddingHorizontal": 0,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "#d1d1d1",
        "borderStyle": "solid",
        "borderTopWidth": 1,
        "flex": 1,
        "flexBasis": 8,
        "minWidth": 8,
      }
    }
  />
</View>
`;

exports[`Divider component tests Subtle Divider 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "center",
      "minHeight": 0,
      "minWidth": 0,
      "paddingHorizontal": 0,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "#f0f0f0",
        "borderStyle": "solid",
        "borderTopWidth": 1,
        "flex": 1,
        "flexBasis": 8,
        "minWidth": 8,
      }
    }
  />
</View>
`;

exports[`Divider component tests Vertical Divider 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "justifyContent": "center",
      "minHeight": 24,
      "minWidth": 0,
      "paddingVertical": 0,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "#e0e0e0",
        "borderLeftWidth": 1,
        "borderStyle": "solid",
        "flex": 1,
        "flexBasis": 8,
        "minHeight": 8,
      }
    }
  />
</View>
`;

exports[`Divider component tests Vertical Divider with Inset 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "justifyContent": "center",
      "minHeight": 24,
      "minWidth": 0,
      "paddingVertical": 16,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "#e0e0e0",
        "borderLeftWidth": 1,
        "borderStyle": "solid",
        "flex": 1,
        "flexBasis": 8,
        "minHeight": 8,
      }
    }
  />
</View>
`;
